packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -55,10 +60,289 @@ enum _ScaffoldSlot {
+/// See also:
+/// 
+///  * [FloatingActionButton], which is a circular button typically shown in the
+///    bottom right corner of the app.
+///  * [FloatingActionButtonPositioner], which the [Scaffold] uses to place the 
+///    [Scaffold.floatingActionButton] within the [Scaffold]'s layout. 
+abstract class FloatingActionButtonAnimator {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const FloatingActionButtonAnimator();
+
+  /// Moves the [FloatingActionButton] by scaling out and in at a new location.
+  /// 
+  /// This is the default [FloatingActionButton] motion animation.
+  static const FloatingActionButtonAnimator scaling = const _ScalingFabMotionAnimator();
+
+  /// Gets the [FloatingActionButton]'s [Offset] based on a [progress].
+  /// 
+  /// [begin] is the [Offset] provided by the previous [FloatingActionButtonPositioner].
+  /// 
+  /// [end] is the [Offset] provided by the new [FloatingActionButtonPositioner].
+  /// 
+  /// [progress] is the current progress of the transition animation.
+  Offset getOffset({@required Offset begin, @required Offset end, @required double progress});
+
+  /// Animates the scale of the [FloatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 1.0.
+  Animation<double> getScaleAnimation({@required Animation<double> parent});
+
+  /// Animates the rotation of [Scaffold.floatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 0.0 or 1.0.
+  /// 
+  /// The animation values are a fraction of a full circle, with 0.0 and 1.0
+  /// corresponding to 0 and 360 degrees, while 0.5 corresponds to 180 degrees.
+  Animation<double> getRotationAnimation({@required Animation<double> parent});
+
+  /// Gets the value to restart a motion animation at if it's interrupted.
+  /// 
+  /// An interruption is triggered if the [Scaffold] is given a new [FloatingActionButtonPositioner]
+  /// while it is still animating a transition between two previous [FloatingActionButtonPositioner]s.
+  /// 
+  /// A sensible default is usually 0.0, which is the same as restarting
+  /// the animation from the beginning.
+  double getAnimationRestart(double previousValue) => 0.0;
+
+  @override
+  String toString() => '$runtimeType';
+}
+
+class _ScalingFabMotionAnimator extends FloatingActionButtonAnimator {
+  const _ScalingFabMotionAnimator();
+
+  @override
+  Offset getOffset({Offset begin, Offset end, double progress}) {
+    if (progress < 0.5) {
+      return begin;
+    } else {
+      return end;
+    }
+  }
+
+  @override
+  Animation<double> getScaleAnimation({Animation<double> parent}) {
+    final Curve curve = const Interval(0.5, 1.0, curve: Curves.ease);
+    return new AnimationMax<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
```

Comment:
maybe add a comment here explaining what this is trying to do? it seems like crazy magic code :-)

Reply:
Done.

Reply:
Using AnimationMax here won't in general work for what you describe in the comment. Suppose we change the curve to a bouncing curve that goes negative. Using AnimationMax will clamp it to zero when it should be negative.

I guess you could use a kind of Animation that adds two child animations, but what you really want is a kind of Animation that defers to one animation for a range of numbers and a second animation for a second range of numbers. That would replace the Interval in your logic here.

Reply:
Done. Please take a look at the _AnimationSwap class I've introduced.

Reply:
Done.  Please take a look at the _AnimationSwap.

I am unsure about making it a part of the public API because I think it's too similar to TrainHoppingAnimation.  That animation doesn't work here, I believe because I can't specify whether to maximize or minimize the animations.

I think long-term it might be better to change the train hopping animation to support this use case.